### PR TITLE
Make nuke script auto-retry upon failure

### PR DESCRIPTION
**Pull request by chuckwondo**
Since terraform commands might fail for various, transient reasons, make
`nuke.sh` automatically retry itself up to a total of 10 attempts before
giving up. This avoids the hassle of having to monitor its execution and
manually rerun it multiple times until all resources are destroyed.

### CODE REVIEW

**Comment by chuckwondo** [comment]
@krisstanton, since the `bin/nuke.sh` script will destroy your entire Cumulus deployment, pull this branch and test the script against your NON-CBA deployment (now that you have your CBA deployment working, you can destroy your old one, which I have done for myself).

Simply make sure your `.env` file points to your "old" deployment, and do the following:

```
git fetch
git checkout auto-retry-nuke
make nuke
```

You will be prompted to enter a specific confirmation phrase, which is to give you a final means of confirming that you are indeed about to "nuke" the correct deployment. If it indicates the correct deployment, enter the indicated phrase, otherwise, simply press enter to abort, and update your `.env` file to point to the right deployment, and run `make nuke` again.

Once the nuke process starts, it will take a couple of hours to fully destroy your deployment, and the script will most likely end up retrying itself multiple times since there will likely be several points at which terraform eventually fails, but these failures are most likely transient failures, and rerunning the script is generally all that's needed to get past a failure, which is the whole point of adding this auto-retry functionality. It should save you a lot of hassle.

Once you successfully nuke your old deployment, not only should you approve this PR, but you should then also mark #118 as Done (I recently created this issue for you).

**Comment by chuckwondo** [comment]
@krisstanton, by the way, once the script has finished, you should be able to confirm that it cleaned up the bulk of your deployment resources by running `make nuke` one last time, which will prompt you again, but after entering the expected phrase, execution should take only a minute or so, and you should see only output similar to the output below.

Further, my previous comment about marking #118 as Done isn't quite right. This will get you most of the way there, but as noted in #118, there are a few more steps in `docs/OPERATING.md` under "Destroying a Deployment" that you must take to clean up a few more resources that are not managed by terraform.

```
++ aws dynamodb list-tables --query '*' --output text
++ grep 'cumulus-chuckwondo-.*Table'
++ tr '\t' '\n'
++ true
+ _dynamo_db_tables=
+ [[ -n '' ]]
++ aws rds describe-db-clusters --query 'DBClusters[?contains(DBClusterIdentifier, '\''cumulus-chuckwondo'\'')].DBClusterIdentifier' --output text
+ _rds_cluster_id=
+ [[ -n '' ]]
+ terraspace down rds-cluster -y
Building .terraspace-cache/us-west-2/chuckwondo/stacks/rds-cluster
Current directory: .terraspace-cache/us-west-2/chuckwondo/stacks/rds-cluster
=> terraform plan -input=false -destroy
Acquiring state lock. This may take a few moments...
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_caller_identity.current: Refreshing state...
module.vpc.data.aws_subnets.ngap_subnets: Refreshing state...
data.aws_region.current: Refreshing state...
module.vpc.data.aws_vpc.ngap_vpc: Refreshing state...
module.rds_cluster.module.provision_database[0].data.aws_iam_policy_document.lambda_assume_role_policy: Refreshing state...
------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
Terraform will perform the following actions:
Plan: 0 to add, 0 to change, 0 to destroy.
------------------------------------------------------------------------
Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
Releasing state lock. This may take a few moments...
Building .terraspace-cache/us-west-2/chuckwondo/stacks/rds-cluster
=> terraform destroy -auto-approve
module.vpc.data.aws_vpc.ngap_vpc: Refreshing state...
data.aws_region.current: Refreshing state...
module.vpc.data.aws_subnets.ngap_subnets: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
module.rds_cluster.module.provision_database[0].data.aws_iam_policy_document.lambda_assume_role_policy: Refreshing state...

Destroy complete! Resources: 0 destroyed.
Releasing state lock. This may take a few moments...
Time took: 10s
```

**Comment by krisstanton** [comment]
Just leaving a quick update here.
I've got `make nuke` running.  It has been going for a while.  It looks like it errored once and auto retried.  It also looks like it is still doing its thing.  
So far so good!

**Comment by krisstanton** [comment]
I am running into a repeated error now.

After about 4 automatic retries (and 4 different errors), the script finally hits this error:

`
Error: error reading Dynamodb Table (cumulus-kris-uat7469-CollectionsTable): ResourceNotFoundException: Requested resource not found: Table: cumulus-kris-uat7469-CollectionsTable not found
`

Got this one on retries 4 - 10.
I tried running
`
make nuke
`
One more time to see if it would resolve it self and is now running into this error everytime.

This looks much like one of those Syncing issues we have solved before.
Would the teraform doctor be able to resolve this one?  
I have not tried using Teraform doctor in the context of destroying a deployment so I don't know if there are some unwanted consequences of that.

**Comment by chuckwondo** [comment]
Terraform doctor is _only_ for deploying, not destroying.

Let me do a bit of digging and get back to you.

**Comment by chuckwondo** [comment]
Your tables are all gone now, so run `make nuke` again and it should get past that.

**Comment by chuckwondo** [comment]
@Kris Stanton, it appears that your Collections table got deleted, but the rest of the tables did not get deleted. Perhaps some error occurred while the script was attempting to loop through the tables and delete them one by one, which then caused a retry.

In that case, the retry attempt sees that there are still tables to be deleted, but the logic is such that the script assumes that the cumulus module was not fully destroyed if there are any tables that still exist. Therefore, when tables still exist, if first tries to destroy the cumulus module before deleting the tables because the cumulus module depends on the tables. 

This ends up being a sort of a catch-22 situation. What we really want here is to _avoid_ another attempt to destroy the cumulus module (because it was already completely destroyed, otherwise none of the tables would have been deleted), and to simply delete the tables.

However, there is no reliable way to know that this is the situation. We would have to hardcode the total number of tables in the script, and only attempt to destroy the cumulus module if _all_ tables still exist. As soon as 1 table is deleted, we know we can skip destroying the cumulus module because no tables will be deleted until the cumulus module is successfully destroyed.

I'm going to see if I can make this more robust, but now that your buckets are gone, you should be able to run `make nuke` again and get past the problem.

**Comment by krisstanton** [comment]
@chuckwondo Thanks for the detailed writeup and explanation.
I did run `make nuke` again and it resulted in the following.

Note, Many of these look like errors but also look like it sort of worked.  (See where it says, `Destroy complete!` but still throws errors.

Here is all the relevant info below.

```
retry 1 - (Actually retry 1 means First actual try)
	Destroy complete! Resources: 0 destroyed.
	....
	An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters.
retry 2
	Destroy complete! Resources: 0 destroyed.
	....
	An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters.
retry 3
	Seahorse::Client::NetworkingError: execution expired
retry 4
	Destroy complete! Resources: 0 destroyed.
	....
	An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters.
retry 5
	Exception Seahorse::Client::NetworkingError: execution expired
retry 6
	Destroy complete! Resources: 0 destroyed.
	....
	An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters.
retry 7
	Destroy complete! Resources: 0 destroyed.
	....
	An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters.
retry 8
	Seahorse::Client::NetworkingError: execution expired
retry 9
	Destroy complete! Resources: 0 destroyed.
	....
	An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters.
retry 10
	Seahorse::Client::NetworkingError: execution expired
```

DETAILS of the 2 types of errors.

Example of the entire run (for retry 6)
// Shows Example of the error: "An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters."
```
++ bin/nuke.sh 6

>>> DANGER! <<<

Proceeding with this operation will
COMPLETELY DESTROY THE FOLLOWING DEPLOYMENT:

    AWS_PROFILE: csdap-cumulus-uat-7469
    TS_ENV:      kris-uat7469

If you wish to proceed, type the following at
the prompt:

    destroy kris-uat7469


++ aws dynamodb list-tables --query '*' --output text
++ tr '\t' '\n'
++ grep 'cumulus-kris-uat7469-.*Table'
++ true
+ _dynamo_db_tables=
+ [[ -n '' ]]
++ aws rds describe-db-clusters --query 'DBClusters[?contains(DBClusterIdentifier, '\''cumulus-kris-uat7469'\'')].DBClusterIdentifier' --output text
+ _rds_cluster_id='cumulus-kris-uat7469-rds-serverless	cumulus-kris-uat7469-rds-serverless-clone'
+ [[ -n cumulus-kris-uat7469-rds-serverless	cumulus-kris-uat7469-rds-serverless-clone ]]
+ terraspace down data-persistence -y
Building .terraspace-cache/us-west-2/kris-uat7469/stacks/data-persistence
Current directory: .terraspace-cache/us-west-2/kris-uat7469/stacks/data-persistence
=> terraform plan -input=false -destroy
Acquiring state lock. This may take a few moments...
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

module.vpc.data.aws_vpc.ngap_vpc: Refreshing state...
module.data_persistence.module.db_migration.data.aws_iam_policy_document.lambda_assume_role_policy: Refreshing state...
data.aws_region.current: Refreshing state...
module.data_persistence.module.db_migration.data.aws_iam_policy_document.db_migration: Refreshing state...
data.aws_caller_identity.current: Refreshing state...
module.vpc.data.aws_subnets.ngap_subnets: Refreshing state...
------------------------------------------------------------------------
An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
Terraform will perform the following actions:
Plan: 0 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
Releasing state lock. This may take a few moments...
Building .terraspace-cache/us-west-2/kris-uat7469/stacks/data-persistence
=> terraform destroy -auto-approve
Acquiring state lock. This may take a few moments...
data.aws_caller_identity.current: Refreshing state...
module.data_persistence.module.db_migration.data.aws_iam_policy_document.db_migration: Refreshing state...
module.vpc.data.aws_subnets.ngap_subnets: Refreshing state...
module.data_persistence.module.db_migration.data.aws_iam_policy_document.lambda_assume_role_policy: Refreshing state...
module.vpc.data.aws_vpc.ngap_vpc: Refreshing state...
data.aws_region.current: Refreshing state...

Destroy complete! Resources: 0 destroyed.
Releasing state lock. This may take a few moments...
Time took: 33s
+ AWS_PAGER=
+ aws rds modify-db-cluster --no-deletion-protection --db-cluster-identifier 'cumulus-kris-uat7469-rds-serverless	cumulus-kris-uat7469-rds-serverless-clone'

An error occurred (InvalidParameterValue) when calling the ModifyDBCluster operation: The input isn't valid. Input can't contain control characters.
++ retry 254 79
++ echo

++ echo -n 'ERROR 254 on line 79.  '
ERROR 254 on line 79.  ++ [[ 6 -ge 10 ]]
++ echo Retrying...
Retrying...
++ echo

++ bin/nuke.sh 7
```

Example of the entire run (for retry 8)
// Shows Example of the error: "Seahorse::Client::NetworkingError: execution expired"
```
++ bin/nuke.sh 8

>>> DANGER! <<<

Proceeding with this operation will
COMPLETELY DESTROY THE FOLLOWING DEPLOYMENT:

    AWS_PROFILE: csdap-cumulus-uat-7469
    TS_ENV:      kris-uat7469

If you wish to proceed, type the following at
the prompt:

    destroy kris-uat7469


++ aws dynamodb list-tables --query '*' --output text
++ tr '\t' '\n'
++ grep 'cumulus-kris-uat7469-.*Table'
++ true
+ _dynamo_db_tables=
+ [[ -n '' ]]
++ aws rds describe-db-clusters --query 'DBClusters[?contains(DBClusterIdentifier, '\''cumulus-kris-uat7469'\'')].DBClusterIdentifier' --output text
+ _rds_cluster_id='cumulus-kris-uat7469-rds-serverless	cumulus-kris-uat7469-rds-serverless-clone'
+ [[ -n cumulus-kris-uat7469-rds-serverless	cumulus-kris-uat7469-rds-serverless-clone ]]
+ terraspace down data-persistence -y
Seahorse::Client::NetworkingError: execution expired
Error evaluating ERB template around line 4 of: /work/config/terraform/backend.tf:
1 terraform {
2   backend "s3" {
3     region         = "<%= expansion(':REGION') %>"
4     bucket         = "<%= bucket('tfstate') %>"
5     key            = "<%= expansion(':ENV/:MOD_NAME/terraform.tfstate') %>"
6     encrypt        = true
7     dynamodb_table = "<%= expansion('cumulus-:ENV-tfstate-locks') %>"
8   }
9 }

Original backtrace (last 8 lines):
/opt/terraspace/embedded/lib/ruby/3.0.0/net/http.rb:987:in `initialize'
/opt/terraspace/embedded/lib/ruby/3.0.0/net/http.rb:987:in `open'
/opt/terraspace/embedded/lib/ruby/3.0.0/net/http.rb:987:in `block in connect'
/opt/terraspace/embedded/lib/ruby/3.0.0/timeout.rb:107:in `timeout'
/opt/terraspace/embedded/lib/ruby/3.0.0/net/http.rb:985:in `connect'
/opt/terraspace/embedded/lib/ruby/3.0.0/net/http.rb:970:in `do_start'
/opt/terraspace/embedded/lib/ruby/3.0.0/net/http.rb:965:in `start'
/opt/terraspace/embedded/lib/ruby/3.0.0/delegate.rb:87:in `method_missing'

Re-run with FULL_BACKTRACE=1 to see all lines
Exception SystemExit: exit
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/render_me_pretty-0.8.4/lib/render_me_pretty/erb.rb:128:in `exit'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/render_me_pretty-0.8.4/lib/render_me_pretty/erb.rb:128:in `print_result'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/render_me_pretty-0.8.4/lib/render_me_pretty/erb.rb:120:in `handle_exception'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/render_me_pretty-0.8.4/lib/render_me_pretty/erb.rb:97:in `rescue in render'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/render_me_pretty-0.8.4/lib/render_me_pretty/erb.rb:85:in `render'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/render_me_pretty-0.8.4/lib/render_me_pretty.rb:11:in `result'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/erb/render.rb:10:in `build'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/strategy/mod/tf.rb:4:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/strategy/mod.rb:6:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/cleaner/backend_change.rb:40:in `fresh_backend'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/cleaner/backend_change.rb:21:in `purge?'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/cleaner/backend_change.rb:10:in `purge'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/cleaner.rb:16:in `backend_change_purge'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/compiler/cleaner.rb:9:in `clean'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/builder.rb:77:in `clean'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/builder.rb:29:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/cli/commander.rb:9:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/cli/plan.rb:43:in `plan_only'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/cli/down.rb:40:in `plan'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/cli/down.rb:19:in `perform'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/cli/down.rb:11:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/cli.rb:95:in `down'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/thor-1.2.1/lib/thor/command.rb:27:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/thor-1.2.1/lib/thor/invocation.rb:127:in `invoke_command'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/thor-1.2.1/lib/thor.rb:392:in `dispatch'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/command.rb:76:in `dispatch'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/thor-1.2.1/lib/thor/base.rb:485:in `start'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/lib/terraspace/cli/concern.rb:65:in `start'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/terraspace-2.2.3/exe/terraspace:7:in `<top (required)>'
/opt/terraspace/embedded/bin/terraspace:25:in `load'
/opt/terraspace/embedded/bin/terraspace:25:in `<top (required)>'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli/exec.rb:58:in `load'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli/exec.rb:58:in `kernel_load'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli/exec.rb:23:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli.rb:486:in `exec'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli.rb:31:in `dispatch'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/cli.rb:25:in `start'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/exe/bundle:48:in `block in <top (required)>'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/lib/bundler/friendly_errors.rb:120:in `with_friendly_errors'
/opt/terraspace/embedded/lib/ruby/gems/3.0.0/gems/bundler-2.3.22/exe/bundle:36:in `<top (required)>'
/opt/terraspace/embedded/bin/bundle:23:in `load'
/opt/terraspace/embedded/bin/bundle:23:in `<main>'
++ retry 1 77
++ echo

++ echo -n 'ERROR 1 on line 77.  '
ERROR 1 on line 77.  ++ [[ 8 -ge 10 ]]
++ echo Retrying...
Retrying...
++ echo

++ bin/nuke.sh 9
```

**Comment by chuckwondo** [comment]
The problem is that you had 2 RDS clusters:

- cumulus-kris-uat7469-rds-serverless
- cumulus-kris-uat7469-rds-serverless-clone (leftover from the DB upgrade -- I had cleaned up my clone during the upgrade process, so I didn't encounter this problem)

The script is written under the assumption you have only 1, so more than 1 breaks the script.

I've updated the script to handle this, and I've also kicked off nuking your deployment again, so do NOT run nuke yourself now.

At this point, for expediency, I'm _not_ going to update the nuke script to handle the bucket issue you encountered earlier, so with this tweak to handle multiple RDS clusters, I'd say you should be able to approve this.

However, do not continue with #118 until I let you know when your deployment is nuked, as following the addition destruction steps in `docs/OPERATING.md` at the same time will likely wreak havoc.

**Comment by krisstanton** [comment]
> I've updated the script to handle this, and I've also kicked off nuking your deployment again, so do NOT run nuke yourself now.

Good catch.
Ok I'll wait until I hear from you that the nuke worked before proceeding on #118 . 
Thanks

**Review by krisstanton — APPROVED** [review]

**Comment by chuckwondo** [comment]
ok, nuke is finished

If you want to continue with #118, either follow the additional instructions in `docs/OPERATING.md` in _this_ branch, or wait until I merge back to main (after you approve this PR), then checkout the `main` branch and pull the latest code, and read the latest `docs/OPERATING.md` there

**Comment by krisstanton** [comment]
> ok, nuke is finished
> 
> If you want to continue with #118, either follow the additional instructions in `docs/OPERATING.md` in _this_ branch, or wait until I merge back to main (after you approve this PR), then checkout the `main` branch and pull the latest code, and read the latest `docs/OPERATING.md` there


I approved so go ahead and merge the branch.  Once I see the merge done, I'll checkout main and proceed.
Thanks again!

**Pull request merged**
